File: intern/145.binary-tree-postorder-traversal.ts
/*
 * @lc app=leetcode id=145 lang=typescript
 *
 * [145] Binary Tree Postorder Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function postorderTraversal(root: TreeNode | null): number[] {
  if (root === null) {
    return [];
  }
  const stack: {
    hasVisited: boolean;
    node: TreeNode;
  }[] = [
    {
      hasVisited: false,
      node: root,
    },
  ];
  const result: number[] = [];

  while (stack.length > 0) {
    const top = stack[stack.length - 1];
    if (top.hasVisited) {
      result.push(top.node.val);
      stack.pop();
    } else {
      top.hasVisited = true;
      if (top.node.right) {
        stack.push({
          hasVisited: false,
          node: top.node.right,
        });
      }
      if (top.node.left) {
        stack.push({
          hasVisited: false,
          node: top.node.left,
        });
      }
    }
  }

  return result;
}
// @lc code=end

File: prev-2024/145.binary-tree-postorder-traversal.cpp
/*
 * @lc app=leetcode id=145 lang=cpp
 *
 * [145] Binary Tree Postorder Traversal
 */
#include <vector>

struct TreeNode {
  int val;
  TreeNode* left;
  TreeNode* right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode* left, TreeNode* right)
      : val(x), left(left), right(right) {}
};

// @lc code=start
class Solution {
 public:
  std::vector<int> postorderTraversal(TreeNode* root) {
    std::vector<int> result;
    recursive(root, result);
    return result;
  }

 private:
  void recursive(TreeNode* root, std::vector<int>& result) {
    if (root == nullptr) {
      return;
    }
    recursive(root->left, result);
    recursive(root->right, result);
    result.push_back(root->val);
  }
};
// @lc code=end
